packages/web/src/routes/session.tsx

```diff
@@ -12,68 +12,105 @@ import {
 import { SessionChannel } from '@/clients/websocket';
 import { CellsProvider, useCells } from '@/components/use-cell';
-import useEffectOnce from '@/components/use-effect-once';
 import { cn } from '@/lib/utils';
 import { useHotkeys } from 'react-hotkeys-hook';
 import InstallPackageModal from '@/components/install-package-modal';
 import { PackageJsonProvider, usePackageJson } from '@/components/use-package-json';
+import { SessionNavbar } from '@/components/navbar';
 import { toast } from 'sonner';
 import { TsConfigProvider } from '@/components/use-tsconfig-json';
 
 async function loader({ params }: LoaderFunctionArgs) {
-  const [{ result: config }, { result: session }] = await Promise.all([
+  const [{ result: config }, { result: srcbooks }, { result: session }] = await Promise.all([
     getConfig(),
+    loadSessions(),
     loadSession({ id: params.id! }),
   ]);
 
-  return { config, session };
+  return { config, srcbooks, session };
 }
 
+type SessionLoaderDataType = {
+  config: SettingsType;
+  srcbooks: Array<SessionType>;
+  session: SessionType;
+};
+
 function SessionPage() {
-  const { session, config } = useLoaderData() as { session: SessionType; config: SettingsType };
+  const { config, srcbooks, session } = useLoaderData() as SessionLoaderDataType;
+
+  // Because we use refs for our state, we need a way to trigger
+  // component re-renders when the ref state changes.
+  //
+  // https://legacy.reactjs.org/docs/hooks-faq.html#is-there-something-like-forceupdate
+  //
+  const [, forceComponentRerender] = useReducer((x) => x + 1, 0);
 
   const channelRef = useRef(SessionChannel.create(session.id));
+  const connectedSessionIdRef = useRef<SessionType['id'] | null>(null);
+  const connectedSessionLanguageRef = useRef<SessionType['language'] | null>(null);
   const channel = channelRef.current;
 
-  useEffectOnce(() => {
+  useEffect(() => {
+    if (connectedSessionIdRef.current === session.id) {
+      return;
+    }
+
+    const oldChannel = channelRef.current;
+
+    // Disconnect from the previously connected session
+    if (connectedSessionIdRef.current) {
+      oldChannel.unsubscribe();
+
+      if (connectedSessionLanguageRef.current === 'typescript') {
+        oldChannel.push('tsserver:stop', { sessionId: session.id });
+      }
+    }
+
+    // Reconnect to the new session
+    channelRef.current = SessionChannel.create(session.id);
+    connectedSessionIdRef.current = session.id;
+    connectedSessionLanguageRef.current = session.language;
+
+    const channel = channelRef.current;
+
     channel.subscribe();
 
     if (session.language === 'typescript') {
       channel.push('tsserver:start', { sessionId: session.id });
     }
 
-    return () => {
-      channel.unsubscribe();
-
-      if (session.language === 'typescript') {
-        channel.push('tsserver:stop', { sessionId: session.id });
-      }
-    };
-  });
+    forceComponentRerender();
+  }, [session.id, session.language, forceComponentRerender]);
```

Comment:
I had to make some signifigant changes to the way that the session page connected to the underlying websocket - the way it previously worked, if a user navigated internally within the app to a new srcbook (ie, by clicking a link, etc), the socket wouldn't disconnect and reconnect, so messages related to a new srcbook would be going to a misconfigured socket, leading to the dev server crashing.

This I think needs a fair amount of testing! As far as I can tell it works locally for me, but I am not sure if I am exercising all the cases.